### PR TITLE
[Merged by Bors] - feat(NumberTheory/Padics/PadicVal): Add `padicValNat_mul_pow_left/right`

### PR DESCRIPTION
**Pull request by gaetanserre**
Add two theorems solving goals of this form: for any primes `p` and `q` such that `p != q`, `padicValNat p (p^n * q^m) = n`

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/Padics/PadicVal.lean [hunk 1]
```diff
@@ -585,6 +585,20 @@ theorem padicValNat_primes {q : ℕ} [hp : Fact p.Prime] [hq : Fact q.Prime] (ne
     (not_congr (Iff.symm (prime_dvd_prime_iff_eq hp.1 hq.1))).mp neq
 #align padic_val_nat_primes padicValNat_primes
 
+theorem padicValNat_prod_pow_left {q : ℕ} [hp : Fact p.Prime] [hq : Fact q.Prime]
+    (n m : ℕ) (neq : p ≠ q) : padicValNat p (p^n * q^m) = n := by
+  have padic_p_qm_eq_0 : padicValNat p (q ^ m) = 0 := by
+    rw [padicValNat.pow _ <| Nat.Prime.ne_zero hq.elim, padicValNat_primes neq, mul_zero]
+  rw [padicValNat.mul (NeZero.ne' (p^n)).symm (NeZero.ne' (q^m)).symm,
+    padicValNat.prime_pow, padic_p_qm_eq_0, add_zero]
```
Comment (BoltonBailey) [inline Mathlib/NumberTheory/Padics/PadicVal.lean:593]:
```suggestion
  rw [padicValNat.mul (NeZero.ne' (p^n)).symm (NeZero.ne' (q^m)).symm, padicValNat.prime_pow, 
    padicValNat.pow _ <| Nat.Prime.ne_zero hq.elim, padicValNat_primes neq, mul_zero, add_zero]
```

Comment (BoltonBailey) [reply Mathlib/NumberTheory/Padics/PadicVal.lean:593]: If you want, you could golf away the `have` like this

**Review by BoltonBailey — APPROVED** [review]
LGTM (see my suggestion below)

File: Mathlib/NumberTheory/Padics/PadicVal.lean (hunk 1 above)
Comment (BoltonBailey) [reply Mathlib/NumberTheory/Padics/PadicVal.lean:593]: Or better yet, you could factor this out as a lemma

```
theorem padicValNat_prime_prime_pow {q : ℕ} [hp : Fact p.Prime] [hq : Fact q.Prime]
    (n : ℕ) (neq : p ≠ q) : padicValNat p (q ^ n) = 0 := by
  rw [padicValNat.pow _ <| Nat.Prime.ne_zero hq.elim, padicValNat_primes neq, mul_zero]

theorem padicValNat_prod_pow_left {q : ℕ} [hp : Fact p.Prime] [hq : Fact q.Prime]
    (n m : ℕ) (neq : p ≠ q) : padicValNat p (p^n * q^m) = n := by
  have padic_p_qm_eq_0 : padicValNat p (q ^ m) = 0 := by exact padicValNat_prime_prime_pow m neq
  rw [padicValNat.mul (NeZero.ne' (p^n)).symm (NeZero.ne' (q^m)).symm,
    padicValNat.prime_pow, padic_p_qm_eq_0, add_zero]
```

Comment (gaetanserre) [reply Mathlib/NumberTheory/Padics/PadicVal.lean:593]: Thank you for your help, I guess the second version is better! I commit your changes

File: Mathlib/NumberTheory/Padics/PadicVal.lean [hunk 2]
```diff
@@ -585,6 +585,20 @@ theorem padicValNat_primes {q : ℕ} [hp : Fact p.Prime] [hq : Fact q.Prime] (ne
     (not_congr (Iff.symm (prime_dvd_prime_iff_eq hp.1 hq.1))).mp neq
 #align padic_val_nat_primes padicValNat_primes
 
+theorem padicValNat_prime_prime_pow {q : ℕ} [hp : Fact p.Prime] [hq : Fact q.Prime]
+    (n : ℕ) (neq : p ≠ q) : padicValNat p (q ^ n) = 0 := by
+  rw [padicValNat.pow _ <| Nat.Prime.ne_zero hq.elim, padicValNat_primes neq, mul_zero]
+
+theorem padicValNat_prod_pow_left {q : ℕ} [hp : Fact p.Prime] [hq : Fact q.Prime]
```
Comment (ocfnash) [inline Mathlib/NumberTheory/Padics/PadicVal.lean:592]: We use `prod` for an indexed product of a family and `mul` for the binary product
```suggestion
theorem padicValNat_mul_pow_left {q : ℕ} [hp : Fact p.Prime] [hq : Fact q.Prime]
```

File: Mathlib/NumberTheory/Padics/PadicVal.lean [hunk 3]
```diff
@@ -585,6 +585,20 @@ theorem padicValNat_primes {q : ℕ} [hp : Fact p.Prime] [hq : Fact q.Prime] (ne
...
+    (n m : ℕ) (neq : p ≠ q) : padicValNat p (p^n * q^m) = n := by
+  rw [padicValNat.mul (NeZero.ne' (p^n)).symm (NeZero.ne' (q^m)).symm,
+    padicValNat.prime_pow, padicValNat_prime_prime_pow m neq, add_zero]
+
+theorem padicValNat_prod_pow_right {q : ℕ} [hp : Fact p.Prime] [hq : Fact q.Prime]
```
Comment (ocfnash) [inline Mathlib/NumberTheory/Padics/PadicVal.lean:597]:
```suggestion
theorem padicValNat_mul_pow_right {q : ℕ} [hp : Fact p.Prime] [hq : Fact q.Prime]
```

File: Mathlib/NumberTheory/Padics/PadicVal.lean [hunk 4]
```diff
@@ -585,6 +585,20 @@ theorem padicValNat_primes {q : ℕ} [hp : Fact p.Prime] [hq : Fact q.Prime] (ne
...
+    (n m : ℕ) (neq : q ≠ p) : padicValNat q (p^n * q^m) = m := by
+  rw [padicValNat.mul (NeZero.ne' (p^n)).symm (NeZero.ne' (q^m)).symm,
```
Comment (ocfnash) [inline Mathlib/NumberTheory/Padics/PadicVal.lean:599]: I'd be tempted to prove this use `mul_comm` and the lemma above but what you have is fine too.

**Review by ocfnash — APPROVED** [review]
I do wonder how often we'll have both `[Fact p.Prime]` and `[Fact q.Prime]` available but in any case there is nothing controversial here.

Thanks!

bors d+

File: Mathlib/NumberTheory/Padics/PadicVal.lean (hunk 4 above)
Comment (gaetanserre) [reply Mathlib/NumberTheory/Padics/PadicVal.lean]: Indeed, that's a better idea. I updated the proof accordingly.

**Pull request closed without merging**